cmd/clusterctl/pkg/client/cluster/components.go

```diff
@@ -49,43 +63,106 @@ func (p *providerComponents) Create(components repository.Components) error {
 		currentR := &unstructured.Unstructured{}
-		currentR.SetGroupVersionKind(r.GroupVersionKind())
+		currentR.SetGroupVersionKind(obj.GroupVersionKind())
 
 		key := client.ObjectKey{
-			Namespace: r.GetNamespace(),
-			Name:      r.GetName(),
+			Namespace: obj.GetNamespace(),
+			Name:      obj.GetName(),
 		}
 		if err = c.Get(ctx, key, currentR); err != nil {
 			if !apierrors.IsNotFound(err) {
 				return errors.Wrapf(err, "failed to get current provider object")
 			}
 
 			//if it does not exists, create the component
-			klog.V(3).Infof("Creating: %s, %s/%s", r.GroupVersionKind(), r.GetNamespace(), r.GetName())
-			if err = c.Create(ctx, &r); err != nil { //nolint
-				return errors.Wrapf(err, "failed to create provider object %s, %s/%s", r.GroupVersionKind(), r.GetNamespace(), r.GetName())
+			klog.V(3).Infof("Creating: %s, %s/%s", obj.GroupVersionKind(), obj.GetNamespace(), obj.GetName())
+			if err = c.Create(ctx, &obj); err != nil {
+				return errors.Wrapf(err, "failed to create provider object %s, %s/%s", obj.GroupVersionKind(), obj.GetNamespace(), obj.GetName())
 			}
 
 			continue
 		}
 
 		// otherwise update the component
-		klog.V(3).Infof("Updating: %s, %s/%s", r.GroupVersionKind(), r.GetNamespace(), r.GetName())
+		klog.V(3).Infof("Updating: %s, %s/%s", obj.GroupVersionKind(), obj.GetNamespace(), obj.GetName())
 
 		// if upgrading an existing component, then use the current resourceVersion for the optimistic lock
-		r.SetResourceVersion(currentR.GetResourceVersion())
-		if err = c.Update(ctx, &r); err != nil { //nolint
-			return errors.Wrapf(err, "failed to update provider object %s, %s/%s", r.GroupVersionKind(), r.GetNamespace(), r.GetName())
+		obj.SetResourceVersion(currentR.GetResourceVersion())
+		if err = c.Update(ctx, &obj); err != nil {
+			return errors.Wrapf(err, "failed to update provider object %s, %s/%s", obj.GroupVersionKind(), obj.GetNamespace(), obj.GetName())
 		}
 	}
 
 	return nil
 }
 
+func (p *providerComponents) Delete(options DeleteOptions) error {
+	// Fetch all the components belonging to a provider.
+	labels := map[string]string{
+		clusterctlv1.ClusterctlProviderLabelName: options.Provider.Name,
+	}
+	resources, err := p.proxy.ListResources(options.Provider.Namespace, labels)
+	if err != nil {
+		return err
+	}
+
+	resourcesToDelete := []unstructured.Unstructured{}
+	namespacesToDelete := sets.NewString()
+	for _, obj := range resources {
+		// If the CRDs should NOT be deleted, skip it;
+		// NB. Skipping CRDs deletion ensures that also the objects of Kind defined in the CRDs Kind are not deleted.
+		if obj.GroupVersionKind().Kind == "CustomResourceDefinition" && !options.ForceDeleteCRD {
+			continue
+		}
+
+		// If the Namespace should NOT be deleted, skip it, otherwise keep track of the namespaces we are deleting;
+		// NB. Skipping Namespaces deletion ensures that also the objects hosted in the namespace but without the "clusterctl.cluster.x-k8s.io/provider" label are not deleted.
```

Comment:
Is this an extra label on top of the one we discussed the other day?

Reply:
same as [1]; to be addressed in a follow-up PR